### PR TITLE
PERF: Fix performance regression with Series statistical ops (#25952)

### LINKED ISSUE #25952: Fix performance regression with Series operations
#25743 caused a performance regression with Series operations (mean, sum, etc).  This can be avoided by moving some of the `if`s around so that the additional check is not performed for those operations.

### PR DESCRIPTION
**Pull request by ArtificialQualia**
- [x] closes #25952
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

#25743 introduced a performance regression in `Series` `MultiIndex` statistical operations.

This happened due to the additional required check for Series groupby operations that already existed for `DataFrame`.  By ensuring that the check needs to be run, this regression can be avoided also potentially slightly speeding up similar `DataFrame` operations.

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
pls create an asv for this or show a current one that is affected
u don’t need a note as this is unreleased code

**Comment by WillAyd** [comment]
Can you post ASV results for this? Also does this impact performance of DataFrame ops?

**Comment by ArtificialQualia** [comment]
Here is the affected ASV tests:

```
       before           after         ratio
     [00c119c5]       [98abd413]
     <master>         <fix-groupby-performance>
-      20.8±0.2ms       18.2±0.5ms     0.88  stat_ops.SeriesMultiIndexOps.time_op(0, 'mad')
-      12.1±0.2ms       9.71±0.2ms     0.80  stat_ops.SeriesMultiIndexOps.time_op(0, 'skew')
-      12.1±0.2ms       9.61±0.2ms     0.79  stat_ops.SeriesMultiIndexOps.time_op(0, 'kurt')
-      8.58±0.2ms      5.96±0.03ms     0.69  stat_ops.SeriesMultiIndexOps.time_op(1, 'sem')
-      8.73±0.1ms       5.79±0.2ms     0.66  stat_ops.SeriesMultiIndexOps.time_op(0, 'sem')
-     8.92±0.08ms      5.92±0.09ms     0.66  stat_ops.SeriesMultiIndexOps.time_op(1, 'median')
-      8.67±0.1ms       5.71±0.1ms     0.66  stat_ops.SeriesMultiIndexOps.time_op(0, 'median')
-      7.23±0.2ms       4.33±0.4ms     0.60  stat_ops.SeriesMultiIndexOps.time_op(1, 'prod')
-      7.67±0.2ms       4.48±0.3ms     0.58  stat_ops.SeriesMultiIndexOps.time_op(0, 'std')
-      7.70±0.2ms      4.39±0.06ms     0.57  stat_ops.SeriesMultiIndexOps.time_op(1, 'var')
-      7.13±0.1ms      3.92±0.03ms     0.55  stat_ops.SeriesMultiIndexOps.time_op(0, 'sum')
-      7.35±0.2ms       4.01±0.1ms     0.55  stat_ops.SeriesMultiIndexOps.time_op(1, 'sum')
-      7.24±0.1ms       3.94±0.2ms     0.54  stat_ops.SeriesMultiIndexOps.time_op(1, 'mean')
-      7.18±0.1ms      3.90±0.05ms     0.54  stat_ops.SeriesMultiIndexOps.time_op(0, 'mean')
-     7.62±0.06ms      4.12±0.04ms     0.54  stat_ops.SeriesMultiIndexOps.time_op(0, 'var')
-      7.15±0.1ms      3.84±0.08ms     0.54  stat_ops.SeriesMultiIndexOps.time_op(0, 'prod')
```

Most `MultiIndex` `DataFrame` operations are improved as well, but only slightly.  Not enough to meet the 10% threshold.

I'll run a full ASV tonight to see if there are any other major affected areas.

**Comment by ArtificialQualia** [comment]
whatsnew line has been removed.

Here are the 'full' ASV results.  I re-ran all the deviant tests and removed those that weren't consistent.  Here are the final results:

```
       before           after         ratio
     [00c119c5]       [98abd413]
     <master>         <fix-groupby-performance>
-      20.5±0.2ms       18.5±0.3ms     0.90  stat_ops.SeriesMultiIndexOps.time_op(0, 'mad')
-     4.49±0.04ms       3.77±0.1ms     0.84  groupby.Size.time_category_size
-      13.9±0.2ms       9.63±0.1ms     0.69  stat_ops.SeriesMultiIndexOps.time_op(0, 'kurt')
-     8.78±0.06ms       5.84±0.1ms     0.67  stat_ops.SeriesMultiIndexOps.time_op(1, 'sem')
-      8.82±0.3ms       5.80±0.1ms     0.66  stat_ops.SeriesMultiIndexOps.time_op(1, 'median')
-      8.94±0.4ms       5.82±0.1ms     0.65  stat_ops.SeriesMultiIndexOps.time_op(0, 'sem')
-      8.70±0.2ms       5.60±0.1ms     0.64  stat_ops.SeriesMultiIndexOps.time_op(0, 'median')
-      7.74±0.1ms      4.75±0.07ms     0.61  stat_ops.SeriesMultiIndexOps.time_op(1, 'std')
-      6.57±0.1ms       3.82±0.1ms     0.58  stat_ops.SeriesMultiIndexOps.time_op(0, 'mean')
-      8.62±0.1ms       4.96±0.2ms     0.58  groupby.TransformBools.time_transform_mean
-      7.64±0.1ms      4.37±0.09ms     0.57  stat_ops.SeriesMultiIndexOps.time_op(0, 'std')
-      7.72±0.2ms      4.39±0.09ms     0.57  stat_ops.SeriesMultiIndexOps.time_op(1, 'var')
-      7.31±0.1ms       4.13±0.1ms     0.57  stat_ops.SeriesMultiIndexOps.time_op(0, 'sum')
-      7.41±0.2ms      4.10±0.08ms     0.55  stat_ops.SeriesMultiIndexOps.time_op(1, 'sum')
-     7.63±0.09ms      4.19±0.09ms     0.55  stat_ops.SeriesMultiIndexOps.time_op(0, 'var')
-      7.47±0.2ms       4.06±0.2ms     0.54  stat_ops.SeriesMultiIndexOps.time_op(1, 'mean')
-      7.35±0.2ms       3.98±0.1ms     0.54  stat_ops.SeriesMultiIndexOps.time_op(1, 'prod')
-      7.27±0.2ms      3.83±0.07ms     0.53  stat_ops.SeriesMultiIndexOps.time_op(0, 'prod')
```

Notably, `groupby.Size.time_category_size` and `groupby.TransformBools.time_transform_mean` are also consistently faster by more than 10%.

File: pandas/core/groupby/grouper.py [hunk 1]
```diff
@@ -520,21 +520,21 @@ def _get_grouper(obj, key=None, axis=0, level=None, sort=True,
...
+        try:
+            if isinstance(obj, DataFrame):
+                all_in_columns_index = all(g in obj.columns or g in
+                                           obj.index.names for g in keys)
+            elif isinstance(obj, Series):
+                all_in_columns_index = all(g in obj.index.names for g in keys)
+            else:
+                all_in_columns_index = False
+        except Exception:
             all_in_columns_index = False
```
Comment (jreback) [inline pandas/core/groupby/grouper.py]: what test actually hits this exception?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/grouper.py (hunk 1 above)
Comment (ArtificialQualia) [reply pandas/core/groupby/grouper.py]: None of them, that was there before the PR that caused this regression.

Also, the
```
else:
    all_in_columns_index = False
````
can probably be removed as no tests hit that now either.  I mentioned that in my last PR, but didn't get any feedback/confirmation that could be removed.

Comment (gfyoung) [reply pandas/core/groupby/grouper.py]: Is it possible to add tests that cover those lines of code?

Comment (ArtificialQualia) [reply pandas/core/groupby/grouper.py]: Just by using public functions?  I don't think it's reachable.  Probably not even reachable using `_get_grouper` directly.  I'll see what I can do.

Comment (jreback) [reply pandas/core/groupby/grouper.py]: if you can remove code and nothing breaks then go for it

File: pandas/core/groupby/grouper.py [hunk 2]
```diff
@@ -520,21 +520,21 @@ def _get_grouper(obj, key=None, axis=0, level=None, sort=True,
...
-        else:
+    if (not any_callable and not any_arraylike and not any_groupers and
+            match_axis_length and level is None):
+        try:
+            if isinstance(obj, DataFrame):
+                all_in_columns_index = all(g in obj.columns or g in
+                                           obj.index.names for g in keys)
+            elif isinstance(obj, Series):
+                all_in_columns_index = all(g in obj.index.names for g in keys)
+            else:
```
Comment (jreback) [inline pandas/core/groupby/grouper.py]: you can probably just set
``all_in_columns_index=False`` before line 523, then remove this else (or is it already being set)?

Comment (ArtificialQualia) [reply pandas/core/groupby/grouper.py]: `all_in_columns_index` is only used in that `if` right below it.

This function expects an `NDFrame`, so I think anything coming into it is going to be a `Series` or `DataFrame` (`Panel` doesn't work with this function for other reasons and is going away anyways).

I can set it to `False` if we want to be safe, but removing the `else` and that `except` block doesn't fail any tests, so it's up to you.  I personally couldn't get it to fail with anything valid I threw at `_getgrouper`

Comment (jreback) [reply pandas/core/groupby/grouper.py]: if you can remove the try/except (and else) then do it

**Comment by ArtificialQualia** [comment]
Anything more needed on this PR?

File: pandas/core/groupby/grouper.py [hunk 3]
```diff
@@ -520,21 +520,16 @@ def _get_grouper(obj, key=None, axis=0, level=None, sort=True,
     any_arraylike = any(isinstance(g, (list, tuple, Series, Index, np.ndarray))
                         for g in keys)
 
-    try:
+    if (not any_callable and not any_arraylike and not any_groupers and
```
Comment (jreback) [inline pandas/core/groupby/grouper.py]: can you add some comments here on what is being checked

**Review by jreback — CHANGES_REQUESTED** [review]
can you show the asv's that improved for this change. I believe the regression is in 0.25, correct? so no whatsnew is then needed. (pls confirm)

**Comment by jreback** [comment]
merge master as well

**Comment by ArtificialQualia** [comment]
Master has been merged

Correct, the change this is fixing was merged in 0.25.  No need for a whatsnew. (see #25743)

Here are the ASV results from a previous comment:

> Here are the 'full' ASV results. I re-ran all the deviant tests and removed those that weren't consistent. Here are the final results:
> 
> ```
>        before           after         ratio
>      [00c119c5]       [98abd413]
>      <master>         <fix-groupby-performance>
> -      20.5±0.2ms       18.5±0.3ms     0.90  stat_ops.SeriesMultiIndexOps.time_op(0, 'mad')
> -     4.49±0.04ms       3.77±0.1ms     0.84  groupby.Size.time_category_size
> -      13.9±0.2ms       9.63±0.1ms     0.69  stat_ops.SeriesMultiIndexOps.time_op(0, 'kurt')
> -     8.78±0.06ms       5.84±0.1ms     0.67  stat_ops.SeriesMultiIndexOps.time_op(1, 'sem')
> -      8.82±0.3ms       5.80±0.1ms     0.66  stat_ops.SeriesMultiIndexOps.time_op(1, 'median')
> -      8.94±0.4ms       5.82±0.1ms     0.65  stat_ops.SeriesMultiIndexOps.time_op(0, 'sem')
> -      8.70±0.2ms       5.60±0.1ms     0.64  stat_ops.SeriesMultiIndexOps.time_op(0, 'median')
> -      7.74±0.1ms      4.75±0.07ms     0.61  stat_ops.SeriesMultiIndexOps.time_op(1, 'std')
> -      6.57±0.1ms       3.82±0.1ms     0.58  stat_ops.SeriesMultiIndexOps.time_op(0, 'mean')
> -      8.62±0.1ms       4.96±0.2ms     0.58  groupby.TransformBools.time_transform_mean
> -      7.64±0.1ms      4.37±0.09ms     0.57  stat_ops.SeriesMultiIndexOps.time_op(0, 'std')
> -      7.72±0.2ms      4.39±0.09ms     0.57  stat_ops.SeriesMultiIndexOps.time_op(1, 'var')
> -      7.31±0.1ms       4.13±0.1ms     0.57  stat_ops.SeriesMultiIndexOps.time_op(0, 'sum')
> -      7.41±0.2ms      4.10±0.08ms     0.55  stat_ops.SeriesMultiIndexOps.time_op(1, 'sum')
> -     7.63±0.09ms      4.19±0.09ms     0.55  stat_ops.SeriesMultiIndexOps.time_op(0, 'var')
> -      7.47±0.2ms       4.06±0.2ms     0.54  stat_ops.SeriesMultiIndexOps.time_op(1, 'mean')
> -      7.35±0.2ms       3.98±0.1ms     0.54  stat_ops.SeriesMultiIndexOps.time_op(1, 'prod')
> -      7.27±0.2ms      3.83±0.07ms     0.53  stat_ops.SeriesMultiIndexOps.time_op(0, 'prod')
> ```
> Notably, `groupby.Size.time_category_size` and `groupby.TransformBools.time_transform_mean` are also consistently faster by more than 10%.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks

**Pull request merged**